### PR TITLE
feat: support lazy loading the lora module for reducing the loading p…

### LINKED ISSUE #433: Improve the latency of `load_batched_adapter_weights`
### Feature request

Currently, every lora layer would be moved from CPU to target device of base model, results in extra 20ms in each layer, finally 500ms ~ 1+s latency overall.

1. first loading into CPU memory
```
def load_module_map(
    ...
    for filename in adapter_filenames:
        adapter_weights.update(load_file(filename))
   ...
```

2. moving to GPU device inside load_batched_adapter_weights
```
lora_a = lora_a.to(base_device, self.dtype)
lora_b = lora_b.to(base_device, self.dtype)
```

### Motivation

Improve the adapter loading performance.

### Your contribution

Yes, I will prepare a PR for review.

### PR DESCRIPTION
**Pull request by thincal**
# What does this PR do?

Fixes #433 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Was this discussed/approved via a Github issue or the discord / slack channel? Please add a link
      to it if that's the case.
- [ ] Did you write any new necessary tests?

## Who can review?

@tgaddair

### CODE REVIEW

**Comment by thincal** [comment]
It seems that caching the handle from safe_open might be a better solution, I will refine it later.

**Comment by thincal** [comment]
> It seems that caching the handle from safe_open might be a better solution, but need to consider the file handle reference management that used by multiple layers, I will refine it later.

Still keep cache the filenames instead of filehandles, since that safe_open needs the device info which differs during loading the lora modules.

**Comment by thincal** [comment]
@tgaddair could you help review this change ?

File: server/lorax_server/utils/adapter.py
Comment (tgaddair) [inline server/lorax_server/utils/adapter.py:169]: Is there any overhead to opening the file for each module we load? Would it be faster to do this once and then make `load_module_weight` a method of a stateful class that maintains a pointer to the open file handle?

**Review by tgaddair — APPROVED** [review]
Looks great @thincal, thanks for the PR, and apologies for the slow review!

I had one question about the file handle, but happy to land this and iterate on it to see if there's any room to further optimize.

File: server/lorax_server/utils/adapter.py
Comment (tgaddair) [reply server/lorax_server/utils/adapter.py:169]: Ah, I see you answered this question above!

**Comment by thincal** [comment]
> Looks great @thincal, thanks for the PR, and apologies for the slow review!
> 
> 
> 
> I had one question about the file handle, but happy to land this and iterate on it to see if there's any room to further optimize.

It is fine to land it firstly, since the safe_open is already lazy behavior and main overhead is about reading out the specific tensor.

**Comment by tgaddair** [comment]
@thincal I noticed there's a failing test:

```
FAILED server/tests/adapters/test_medusa.py::test_batched_medusa_weights - safetensors_rust.SafetensorError: device cpu is invalid
```

Would you be able to take a look before we merge? We should be good to go once that's resolved.

**Comment by thincal** [comment]
> @thincal I noticed there's a failing test:
> 
> ```
> FAILED server/tests/adapters/test_medusa.py::test_batched_medusa_weights - safetensors_rust.SafetensorError: device cpu is invalid
> ```
> 
> Would you be able to take a look before we merge? We should be good to go once that's resolved.

OK, I will finish it today, thanks.

**Comment by thincal** [comment]
@tgaddair fix passed for `server/tests/adapters/test_medusa.py::test_batched_medusa_weights`, the remained errors seem  related with repo access failure, could you help have a check ?

**Comment by thincal** [comment]
@tgaddair ping, sorry for my late response. Could you help review the revised commit? Thanks.

**Comment by tgaddair** [comment]
Hey @thincal, very sorry for the delay here. Let me take a look now.

**Review by tgaddair — APPROVED** [review]
Thanks for this PR and again really sorry for the delay in getting it landed! Sanity checks look good on my side. May run some more benchmarking later, but good to go ahead and land.

**Comment by thincal** [comment]
@tgaddair it's glad to see this change be merged, and thanks for your support :)

**Pull request merged**
